syntax = "proto3";

package product;

service ProductService {

    /* 
    * Checks the availability of stock for a given product.
    * Accepts a ProductRequest containing product details and returns a ProductResponse
    * indicating the stock status.
    */
    rpc CheckStock (ProductRequest) returns (ProductResponse);

    /*
    * Retrieves detailed information about a specific product.
    * Accepts a ProductIdRequest containing the product ID and returns a ProductDetailResponse
    * with comprehensive product details.
    */
    rpc GetProduct (ProductIdRequest) returns (ProductDetailResponse);

    /*
     * Reduces the stock quantity for a list of products.
     *
     * This RPC method accepts a `ReduceStockRequest` containing a list of product IDs 
     * and their respective quantities to reduce. It returns a `ReduceStockResponse` 
     * indicating whether the operation was successful or not.
     */
    rpc ReduceStock (ReduceStockRequest) returns (ReduceStockResponse);
}

message ProductRequest {
    string id = 1;
    int32 quantity = 2;
}

message ProductIdRequest {
    string id = 1;
}

message ProductResponse {
    bool available = 1;
    double price = 2;
    string vendor = 3;
    string name = 4;
    string slug = 5;
}

message ProductDetailResponse {
    string id = 1;
    string name = 2;
    string slug = 3;
    double price = 4;
}

message ReduceStockItem {
    string id = 1;
    int32 quantity = 2;
}

message ReduceStockRequest {
    repeated ReduceStockItem items = 1;  // List of products
}

message ReduceStockResponse {
    bool success = 1;
    repeated string messages = 2;  // List of messages per item
}
